File: DAA_Project/msis_dp.py
"""Dynamic Programming solution for Maximum Sum Increasing Subsequence (MSIS)."""

from typing import List, Tuple


def maximum_sum_increasing_subsequence(arr: List[int]) -> Tuple[int, List[int]]:
    """
    Compute the maximum sum increasing subsequence and reconstruct the subsequence.

    DP state:
    dp[i] = maximum sum of an increasing subsequence ending at index i.

    Recurrence:
    dp[i] = max(dp[i], dp[j] + arr[i]) for all j < i with arr[j] < arr[i]

    Why this differs from LIS:
    - LIS optimizes subsequence length.
    - MSIS optimizes subsequence sum.

    Time Complexity: O(n^2)
    Space Complexity: O(n)
    """
    if not arr:
        return 0, []

    n = len(arr)
    dp = arr[:]  # Base case: subsequence containing only arr[i]
    parent = [-1] * n  # For reconstruction

    for i in range(n):
        for j in range(i):
            if arr[j] < arr[i] and dp[j] + arr[i] > dp[i]:
                dp[i] = dp[j] + arr[i]
                parent[i] = j

    max_sum = max(dp)
    end_index = dp.index(max_sum)

    # Backtrack to construct subsequence.
    subsequence: List[int] = []
    while end_index != -1:
        subsequence.append(arr[end_index])
        end_index = parent[end_index]
    subsequence.reverse()

    return max_sum, subsequence

File: msis_dp.c
#include "msis_dp.h"

#include <stdlib.h>

int maximum_sum_increasing_subsequence(
    const int *arr,
    int n,
    int **out_subsequence,
    int *out_subsequence_length
) {
    int i;
    int j;
    int max_sum;
    int end_index;
    int *dp;
    int *parent;
    int *subsequence;
    int length = 0;
    int walk;

    if (out_subsequence == NULL || out_subsequence_length == NULL) {
        return 0;
    }

    *out_subsequence = NULL;
    *out_subsequence_length = 0;

    if (arr == NULL || n <= 0) {
        return 0;
    }

    dp = (int *)malloc((size_t)n * sizeof(int));
    parent = (int *)malloc((size_t)n * sizeof(int));
    if (dp == NULL || parent == NULL) {
        free(dp);
        free(parent);
        return 0;
    }

    for (i = 0; i < n; i++) {
        dp[i] = arr[i];
        parent[i] = -1;
    }

    for (i = 0; i < n; i++) {
        for (j = 0; j < i; j++) {
            if (arr[j] < arr[i] && dp[j] + arr[i] > dp[i]) {
                dp[i] = dp[j] + arr[i];
                parent[i] = j;
            }
        }
    }

    max_sum = dp[0];
    end_index = 0;
    for (i = 1; i < n; i++) {
        if (dp[i] > max_sum) {
            max_sum = dp[i];
            end_index = i;
        }
    }

    walk = end_index;
    while (walk != -1) {
        length++;
        walk = parent[walk];
    }

    subsequence = (int *)malloc((size_t)length * sizeof(int));
    if (subsequence == NULL) {
        free(dp);
        free(parent);
        return 0;
    }

    walk = end_index;
    for (i = length - 1; i >= 0; i--) {
        subsequence[i] = arr[walk];
        walk = parent[walk];
    }

    *out_subsequence = subsequence;
    *out_subsequence_length = length;

    free(dp);
    free(parent);
    return max_sum;
}